### PR TITLE
Float16/32 inconsistent compact NaN printing #23248

### LINKED ISSUE #23248: Float16/32 inconsistent compact NaN printing
While writing tests for #23244, I encountered this inconsistency:

```jl
julia> AbstractFloat[NaN16, NaN32, NaN64]
3-element Array{AbstractFloat,1}:
   NaN
 NaN32
   NaN
```

For compact printing, either `NaN16` should print as "NaN16" or `NaN32` should print as "NaN". Both changes are fairly straightforward, to implement the former, just change the `false` on this line to `true`:

https://github.com/JuliaLang/julia/blob/131807cc574bc8452b463247d4903d2eb6cfec6b/base/grisu/grisu.jl#L127

To implement the latter, just change the `true` on this line to `x isa Float32`:

https://github.com/JuliaLang/julia/blob/131807cc574bc8452b463247d4903d2eb6cfec6b/base/grisu/grisu.jl#L119

In either case, tests should be added to check the printing behavior of `NaN` in compact context. This is quite straightforward so I'm labeling it as an intro issue.

### PR DESCRIPTION
**Pull request by Vaibhavdixit02**
fix  #23248

### CODE REVIEW

File: test/numbers.jl [hunk 1]
```diff
@@ -3023,3 +3023,10 @@ end
         end
     end
 end
+
+@testset "compact NaN printing" begin
+    @test sprint(io->show(IOContext(io, :compact => true), NaN16)) == "NaN"
+    @test_broken sprint(io->show(IOContext(io, :compact => true), NaN32)) == "NaN"
```
Comment (fredrikekre) [inline test/numbers.jl]: This one should be `@test`, it is what this PR fixes :)

**Comment by Vaibhavdixit02** [comment]
@StefanKarpinski there seems to be some problem with the change.

File: base/grisu/grisu.jl [hunk 2]
```diff
@@ -116,7 +116,7 @@ end
 
 function Base.show(io::IO, x::Union{Float64,Float32})
     if get(io, :compact, false)
-        _show(io, x, PRECISION, 6, true, true)
+        _show(io, x, PRECISION, 6, isa(x,Float32), true)
```
Comment (StefanKarpinski) [inline base/grisu/grisu.jl]: Sorry, I got this condition backwards, which is why the test is failing. Should be `x isa Float64`.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

File: base/grisu/grisu.jl (hunk 2 above)
Comment (Vaibhavdixit02) [inline base/grisu/grisu.jl]: Yeah I was thinking the same thing!

**Comment by StefanKarpinski** [comment]
Thanks for the fix, @Vaibhavdixit02!

**Pull request merged**